### ISSUE TITLE
Google Colab support

### ISSUE THREAD

**Issue by filippo82** [bclavie/RAGatouille#13]
Hi there,

do you have any ideas/clues for the issue with Google Colab? If yes, I could look into this?

**Comment by bclavie** [comment]
Hey Filippo! Also tagging @okhat as it's something we've been discussing and that we want to fix. The issue seems to be down to us using multiprocessing/forking, which hangs forever in certain environments.

The fix, or workaround, would be to simply not fork when there is a single GPU, which would cover most of the situations where the hanging is noticeable (assuming that it's very uncommon to train on multi-GPU in colab/other non-jupyter notebooks envs). I think upstreaming it to the main ColBERT repo would be ideal: https://github.com/stanford-futuredata/ColBERT . You can check out any of the main files (Indexer.py, Trainer.py, etc...) to get an idea of how the process pool is handled at the moment.

**Comment by filippo82** [comment]
Would you have a sample notebook I could use to test the issue? Not that I am an expert in this ... but I could have a look at it at least.

**Comment by bclavie** [comment]
The example training notebook is a good one! If you run it on Colab, it will hang once you get to the `train()` call (when the processes fork). The goal would be that unless n_gpus is explicitly >1, it we shouldn't be using multiprocessing, which will eliminate the problem.

**Comment by filippo82** [comment]
A quick update after an initial run of the example training notebook:

* I uploaded the `02-basic_training.ipynb` to Google Colab
* I selected a T4 GPU
* I installed RAGatouille with `pip install ragatouille`
* I restarted the session

The notebook does **NOT** hang while executing the cell with `train()` but it completes after 20s seconds with this output:

```
#> Starting...
#> Joined...
```

**Comment by filippo82** [comment]
@bclavie let me know if I am missing anything ... it seems to be working fine for me with a T4 GPU on Google Colab.

**Comment by bclavie** [comment]
Hey @filippo82, this is strange, I'm still getting hangups 🤔. I will be trying to have a deeper look at some point to figure out exactly what's causing the hangups!

```bash
#> Starting...
#> Joined...
```

This output is also quite strange, as I'd expect it to get the normal main process prints (i.e. the training config and the loss at each training step)

Out of curiosity, does Indexing run fine on your end too?

**Comment by filippo82** [comment]
Hi @bclavie, if you see the `#> Joined...` message, then it means it finished the execution of that cell. You can find the additional logs by clicking on Runtime -> View runtime logs:

![Screenshot 2024-01-09 at 15 44 37](https://github.com/bclavie/RAGatouille/assets/5595364/2cae3e5a-7f53-42e2-a869-b7ec3c57bb48)

**Comment by bclavie** [comment]
Oh I didn't realise it was a truncated output you were sharing, I thought you meant that was the whole output!

Colab still hangs when I try it, and it still hangs on Windows (though less important a problem), so there's something fishy going on 🤔 I'll try and dig to figure it out
